File: engine/src/main/java/de/hanno/hpengine/engine/model/loader/md5/AnimationState.java
package de.hanno.hpengine.engine.model.loader.md5;

public class AnimationState {
    private final float fps;
    private final float spf;
    private int maxFrames;
    private int currentFrame;
    private int lastFrame;
    private boolean hasUpdated = false;
    private float currentSeconds = 0;

    public AnimationState(int maxFrames, float fps) {
        this.maxFrames = maxFrames;
        this.fps = fps;
        this.spf = 1f/fps;
    }

    public void nextFrame() {
        int nextFrame = currentFrame + 1;
        if (nextFrame > maxFrames - 1) {
            currentFrame = 0;
        } else {
            currentFrame = nextFrame;
        }
        setHasUpdated(true);
    }

    public boolean isHasUpdated() {
        hasUpdated = !(lastFrame == currentFrame);
        return hasUpdated;
    }

    public void setHasUpdated(boolean hasUpdated) {
        if (this.hasUpdated) {
            lastFrame = currentFrame;
        }
        this.hasUpdated = hasUpdated;
    }

    public int getCurrentFrame() {
        return currentFrame;
    }

    public void update(float seconds) {
        currentSeconds += seconds;
        if(currentSeconds > spf) {
            currentSeconds -= spf;
            nextFrame();
        }
    }
}

File: engine/src/main/java/de/hanno/hpengine/engine/model/AnimationState.kt
package de.hanno.hpengine.engine.model

class Animation(val name: String,
                val frames: List<AnimatedFrame>,
                val duration: Float,
                var fps: Float) {

    private val maxFrames = frames.size
    private val spf: Float
        get() = 1f / fps

    var currentFrame = 0
        private set

    private var lastFrame = 0

    var hasUpdated = false
        get() {
            field = lastFrame != currentFrame
            return field
        }
        set(hasUpdated) {
            if (this.hasUpdated) {
                lastFrame = currentFrame
            }
            field = hasUpdated
        }

    private var currentSeconds = 0f
    fun nextFrame() {
        val nextFrame = currentFrame + 1
        currentFrame = if (nextFrame > maxFrames - 1) {
            0
        } else {
            nextFrame
        }
        hasUpdated = true
    }

    fun update(seconds: Float) {
        currentSeconds += seconds
        if (currentSeconds > spf) {
            currentSeconds -= spf
            nextFrame()
        }
    }

}